File: examples/langgraph-picket/_plumbing_check.mjs
// Layer test: the same picket-mcp wiring exercised with a RAW MCP client — no
// LangGraph. If this passes and the graph fails, the problem is in the graph;
// if this fails, the problem is in the picket wiring. Run: node _plumbing_check.mjs
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';

const here = (p) => fileURLToPath(new URL(p, import.meta.url));
const PICKET_MCP = process.env.PICKET_MCP || here('../../bin/picket-mcp.mjs');
const PAGE = readFileSync(here('./hostile-invoice.html'), 'utf8');

const client = new Client({ name: 'plumbing-check', version: '0.0.0' });
await client.connect(new StdioClientTransport({
  command: process.execPath,
  args: [PICKET_MCP],
  env: { ...process.env, PICKET_ALLOWLIST: 'acme.example', PICKET_CDP: '' },
  stderr: 'pipe',
}));

const { tools } = await client.listTools();
const names = tools.map((t) => t.name).sort();
console.log('tools:', names);

const r = await client.callTool({ name: 'picket_observe', arguments: { html: PAGE, task: 'process the acme invoice' } });
const text = r.content.map((b) => b.text).join('\n');
console.log('observe verdict line:', text.split('\n')[0]);

const ok =
  names.join(',') === 'picket_gate,picket_login,picket_observe' &&
  /BLOCK/.test(text) &&
  !/exfil\.evil\.example/.test(text) &&
  /Invoice #7/.test(text);
await client.close();
console.log(ok ? 'PLUMBING_PASS' : 'PLUMBING_FAIL');
process.exit(ok ? 0 : 1);

File: examples/autogen-picket/_plumbing_check.py
"""Layer test: the same picket-mcp wiring exercised with a RAW MCP client - no
CrewAI. If this passes and the flow fails, the problem is in the flow; if this
fails, the problem is in the picket wiring. Run: python _plumbing_check.py"""
from __future__ import annotations

import asyncio
import os

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client

HERE = os.path.dirname(os.path.abspath(__file__))
PICKET_MCP = os.environ.get("PICKET_MCP", os.path.join(HERE, "..", "..", "bin", "picket-mcp.mjs"))
with open(os.path.join(HERE, "hostile-invoice.html"), encoding="utf-8") as f:
    PAGE = f.read()


async def run() -> bool:
    params = StdioServerParameters(
        command="node",
        args=[PICKET_MCP],
        env={**os.environ, "PICKET_ALLOWLIST": "acme.example", "PICKET_CDP": ""},
    )
    async with stdio_client(params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()
            tools = sorted(t.name for t in (await session.list_tools()).tools)
            print("tools:", tools)
            res = await session.call_tool(
                "picket_observe", {"html": PAGE, "task": "process the acme invoice"}
            )
            text = "\n".join(b.text for b in res.content if getattr(b, "text", None))
            print("observe verdict line:", text.splitlines()[0])
            return (
                tools == ["picket_gate", "picket_login", "picket_observe"]
                and "BLOCK" in text
                and "exfil.evil.example" not in text
                and "Invoice #7" in text
            )


if __name__ == "__main__":
    ok = asyncio.run(run())
    print("PLUMBING_PASS" if ok else "PLUMBING_FAIL")
    raise SystemExit(0 if ok else 1)
